adding scala micro benchmarks

tanmaykm opened the pull request:
All benchmark code are in `perf.scala`.
Uses `breeze` (https://github.com/scalanlp/breeze) for linear algebra.
Requires `sbt` to be installed; `sbt` pulls in all other required dependencies.

tanmaykm committed:
adding scala micro benchmarks

All benchmark code are in `perf.scala`.
Uses `breeze` (https://github.com/scalanlp/breeze) for linear algebra.
Requires `sbt` to be installed; `sbt` pulls in all other required dependencies.

tanmaykm commented:
cc: @ViralBShah

tkelman commented:
Coop. What do the relative timings look like?

tanmaykm commented:
On my laptop:

````
scala,fib,0.000156548
scala,parse_int,0.000007552
scala,mandel,0.110920130
scala,quicksort,0.426432623
scala,pi_sum,38.649330000
scala,rand_mat_mul,138.042774000
scala,rand_mat_stat,38.137788000
scala,printfd,50.158738000
````

vs

````
julia,fib,0.051502,0.126824,0.051874,0.002397
julia,parse_int,0.160278,1.663407,0.169986,0.057365
julia,mandel,0.143264,0.282093,0.151010,0.005552
julia,quicksort,0.337883,0.493786,0.353625,0.008989
julia,pi_sum,38.638445,39.238125,38.766917,0.114253
julia,rand_mat_stat,8.894564,10.809974,9.552615,0.385211
julia,rand_mat_mul,54.301580,66.860332,58.869989,4.330207
julia,printfd,18.070467,21.318644,18.376827,0.518694
````

Not sure if something's getting optimized away in `fib` and `parse_int`. Can't make out much from `javap` disassembly.

ViralBShah commented:
How to deal with those cases? I guess this is ok to merge for now, and over time, scala can also enter the benchmarks picture and table.

hayd commented:
scala has TCO so fib could be legit...

johnmyleswhite commented on test/perf/micro/scala/src/main/scala/perf.scala:
```diff
@@ -0,0 +1,242 @@
+
+  // time fib
+  def fib(n:Int):Int = {
+    def fib_tail(n: Int, a:Int, b:Int): Int = n match {
```
This isn't the algorithm we're using for the other benchmarks.

johnmyleswhite commented:
I'd argue that the TCO issue shouldn't come up at all because the Julia implementation doesn't employ a form of tail call recursion, only this Scala code does.

In particular, the algorithm being used in the Scala code has a very different computational complexity than the algorithm we're using to benchmark Julia. The Scala code is essentially equivalent to an iterative algorithm; the use of recursion is purely a syntactic consideration.

hayd commented:
The following is a direct, working, port.
```
def fib(n:Int):Int = if (n < 2) n else fib(n-1) + fib(n-2)
```

---

Aside why do we take the minimum time, rather than the average in some tests (but not others)?

tanmaykm commented:
fib on my laptop without TCO:

````
scala,fib,0.028770707
````

tanmaykm commented:
Updated `fib` as suggested, corrected time values for `parseint` (single loop's time was being divided by number of loops, instead of the cumulative time).

````
scala,fib,0.028404098
scala,parse_int,0.002423477
scala,mandel,0.116083050
scala,quicksort,0.424234483
scala,pi_sum,40.025806000
scala,rand_mat_stat,37.591107000
scala,rand_mat_mul,141.924834000
scala,printfd,51.059306000
````

ViralBShah commented:
We should uniformly use min time or average time in all benchmarks. @hayd Would be great if you can provide a PR. We can then update the tables and such and also add scala to the mix.

StefanKarpinski commented:
It should be the minimum since none of these benchmarks should involve significant GC.

ViralBShah commented:
Merging for now, but I am sure will keep seeing improvements.

hayd commented:
I'll have a look, perhaps it would be good to use the same timing algorithm/abstraction in every language.
